Add Multiple Approaches to Approximate BCS Games and Discuss Group-Theoretic Criteria

Wilson123d opened the pull request:
This pull request provides two different ways to approximate the commuting operator value of a binary constraint system (BCS) game in toqito and links the results to the group-theoretic condition from [arXiv:1606.02278](https://arxiv.org/abs/1606.02278) about the existence of a central element J != e. Actually, J is presented for -1. 
Method A: from_bcs_game
We define a function bcs(M, b) that turns each row of  M including a factor or +1 / -1 arises from (-1)^b[i] (for ith row), then the implementation of built tool NonlocalGame.from_bcs_game from Toqito is applied directly. The theory behind this is we could suppose the solution group is generated by g, for each row of M, we could write summation over j of (Mijxj) = b[i] into product form of gi^Mji where we set gi = (-1)^(xi), this is from definition 3 in the paper. 

Method B: Manual Construction of NonlocalGame
We explicitly create the probability distribution over (Alice question, Bob question) pairs and build the predicate tensor that encodes the valid answer pairs. This approach offers more transparency into how the constraints are mapped to a nonlocal game.
Both methods yield an upper bound on the commuting measurement value for the BCS game at the specified NPA level.

Wilson123d commented:
> @Wilson123d it seems you added files unintentionally from the cycle benchmarking issue. We want to remove those from this diff. It's also not clear to me how or where in your code that you are checking for whether the strategy is perfect or not. It seems like you are just computing the commuting measurement operator value of a BCS (which we already have the functionality to do).

Thanks for your message. I have removed cycle_benchmarking.ipynb. what I thought is I encode the binary constraint system Mx=b into a nonlocal game by setting up the inputs (questions) for Alice and Bob, and the corresponding winning conditions (constraints). These winning conditions reflect the equations from Mx=b. Then obtained an upper bound on the maximum winning probability that quantum players can achieve when their measurement operators commute. If the computed value is 1, it means that there is a strategy such that the players can win the game with probability 1. According to Theorem 4 in the paper, a winning probability of 1 is equivalent to the existence of a perfect commuting‑operator quantum strategy (which in turn means there exists an operator solution for Mx=b and the associated solution group has a nontrivial element 𝐽.

vprusso commented:
@Wilson123d you still have `cycle_benchmarking.py` which you need to remove. 

I would imagine that the resolution of this feature would include a function `has_perfect_commuting_measurement_strategy` (or similar) that is a property acting on a BCS game. This doesn't appear to be what you've provided, and what you're committing instead seem to be one-off scripts that don't appear to be integrated within toqito, have tests, conform to the style, etc. 

I encourage you to refer to the contribution guide for more information on how to contribute to the toqito project:
https://toqito.readthedocs.io/en/latest/contributing.html#contrib-guide-reference-label

purva-thakre commented on BcS.py:
```diff
@@ -0,0 +1,152 @@
+            game._perfect = True
+            return game
+
+# Unit Tests
```
@Wilson123d As @vprusso suggested, it might be worth reading through the contributing guide in the documentation. 

Unit tests are added in a separate file. Please take a look at how `toqito` is organized.

purva-thakre commented on test_BCS.py:
```diff
@@ -0,0 +1,52 @@
+print("Starting tests in test_BCS.py...")
+
+import unittest
```
@Wilson123d We use pytest not unittest

purva-thakre commented on BCS.py:
```diff
@@ -0,0 +1,104 @@
+            return False
+    return True
+
+class NonlocalGame:
```
There already exists a class with this name in toqito.

vprusso replied:
This is defined in the `nonlocal_games/` module. Is there are reason you are defining it in this file?

vprusso commented on BCS.py:
```diff
@@ -0,0 +1,104 @@
+            return 1.0 if self._perfect else 0
+        return 0
+
+    def has_perfect_commuting_measurement_strategy(self, tol: float = 1e-6) -> bool:
+        comm_value = self.commuting_measurement_value_upper_bound(k=2)
+        return np.isclose(comm_value, 1.0, atol=tol)
```
This doesn't seem to be doing what we want. We can already compute the commuting measurement value numerically. We want a method to let us know if a perfect strategy exists or not. In doing so, we specifically do not want to invoke the upper bound function, we instead want to use the group theoretic approach that is outlined in the issue.

vprusso commented on test_BCS.py:
```diff
@@ -0,0 +1,52 @@
+print("Starting tests in test_BCS.py...")
```
Please refer to the contribution guide for how to write and where to put tests.

vprusso commented on toqito/nonlocal_games/bcs.py:
```diff
@@ -0,0 +1,119 @@
+    # This is to check if the nontrivial J exists, consistent with lemma 9.
+    return has_cycle
+
+class BCSNonlocalGame(NonlocalGame):
```
I wonder if defining this class (and also the `from_bcs_game` function is necessary. We already have such a method that is defined for any `NonlocalGame` object. 

Presumably, you could have something that looks like:

```py
from toqito.nonlocal_games import NonlocalGame

bcs_constraints = ...
game = NonlocalGame.from_bcs_game(bcs_constraints)

# Your function which would return either `True` or `False` depending on 
# whether it has a perfect strategy or not.
game.has_perfect_commuting_measurement_strategy()
```

Does the above make sense? It seems redundant to me to be redefining things that already exist in the `nonlocal_game.py` module.

Wilson123d replied:
Thank you for your feedback! I have removed the BCSNonlocalGame class, as it is indeed unnecessary. To align with the existing NonlocalGame class from Toqito, I introduced functions to convert the input (M,b) into the required constraints format.

However, to fully integrate the functionality, we may need to modify NonlocalGame in Toqito by adding the _perfect attribute and defining a has_perfect_commuting_measurement_strategy() method that simply returns self._perfect. Additionally, within from_bcs_game, we might insert perfect = check_perfect_commuting_strategy(M, b) and update game._perfect accordingly—similar to what was previously handled in BCSNonlocalGame.

The updated tests all pass using check_perfect_commuting_strategy(M: np.ndarray, b: np.ndarray) -> bool. However, when using NonlocalGame directly from Toqito, an error occurs due to the missing _perfect attribute and has_perfect_commuting_measurement_strategy() method.

Lastly, I ensured that the constraints are stored in the solution group formalism for consistency.

vprusso replied:
Thanks for the update @Wilson123d 

You may also want to take a peek at how to format comments on GitHub using markdown as they can sometimes be hard to follow:
https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

purva-thakre commented on toqito/nonlocal_games/test_bcs.py:
```diff
@@ -0,0 +1,51 @@
+    M = np.array([[1, 1], [1, 1]], dtype=int)
+    b = np.array([0, 1], dtype=int)
+    constraints = bcs(M, b)
+    assert check_perfect_commuting_strategy(M, b) is False
```
```suggestion
    assert not check_perfect_commuting_strategy(M, b)
```
You don't really need `is True` or `is False` at the end.

Wilson123d replied:
Thanks for your suggestion. I have modified it.

purva-thakre commented on toqito/nonlocal_games/test_bcs.py:
```diff
@@ -0,0 +1,51 @@
+import numpy as np
```
Check where this file is placed compared to other tests files in `nonlocal_games`. 

Please read the contributing guide. https://toqito.readthedocs.io/en/latest/contributing.html

You need to import your main function into the init file too.

Wilson123d replied:
Thanks for your suggestion. I have moved test_binary_constraint_games.py to the tests files.

Wilson123d commented on toqito/nonlocal_games/bcs.py:
```diff
@@ -0,0 +1,126 @@
+    parity=[]
+    for i in b:
+        parity.append(int(i))
+    return row, parity
```
Thanks for your suggestion. Your code is completely correct. I have simplified it.

Wilson123d commented on toqito/nonlocal_games/bcs.py:
```diff
@@ -0,0 +1,126 @@
+import numpy as np
```
Thanks for your suggestion. I have renamed both files.

Wilson123d commented on toqito/nonlocal_games/bcs.py:
```diff
@@ -0,0 +1,126 @@
+        parity.append(int(i))
+    return row, parity
+
+def check_perfect_commuting_strategy(M: np.ndarray, b: np.ndarray) -> bool:
```
Thanks for your suggestion. I  tried to simplified it.

Wilson123d commented on toqito/nonlocal_games/test_bcs.py:
```diff
@@ -0,0 +1,51 @@
+import numpy as np
+from toqito.nonlocal_games.nonlocal_game import NonlocalGame
+from BCS import bcs, check_perfect_commuting_strategy, nonlocal_game_from_constraints
```
Thanks for your suggestion, yes, the error should be the name of file BCS. I used capital letter but github saved it as bcs, it should be fixed now.

Wilson123d commented on toqito/nonlocal_games/test_bcs.py:
```diff
@@ -0,0 +1,51 @@
+    constraints = bcs(M, b)
+    assert check_perfect_commuting_strategy(M, b) is True
+    #game = nonlocal_game_from_constraints(constraints)
+    #assert game.has_perfect_commuting_measurement_strategy() is True
```
> Yes, that line was commented out intentionally. The reason is that calling `has_perfect_commuting_measurement_strategy()` would raise an error because this method is not defined in the `NonlocalGame` class in Toqito.  
>  
> Previously, I created a subclass called `BCSNonlocalGame` specifically to run these tests. That subclass included both the `has_perfect_commuting_measurement_strategy()` method and the `_perfect` attribute to support this functionality.

purva-thakre commented on toqito/nonlocal_games/binary_constraint_system_game.py:
```diff
@@ -0,0 +1,127 @@
+    Each row in M is converted into a bitmask by summing powers of 2 for columns
+    where M[i, j] == 1. The parity bits are taken directly from b.
+    
+    Returns:
```
We have a different style for specifying params. Please take a look at the contributing guide in the documentation. 
https://toqito.readthedocs.io/en/latest/contributing.html#contrib-guide-reference-label

vprusso commented on toqito/nonlocal_games/binary_constraint_system_game.py:
```diff
@@ -0,0 +1,127 @@
+import networkx as nx
+from toqito.nonlocal_games.nonlocal_game import NonlocalGame
+
+def bcs(M, b):
```
I think this function should be renamed to something other than `bcs`. Is there not some way to rectify the way in which a BCS game is constructed from the constraint list within the `nonlocal_game.py` file and this function?

vprusso commented on toqito/nonlocal_games/binary_constraint_system_game.py:
```diff
@@ -0,0 +1,127 @@
+    where M[i, j] == 1. The parity bits are taken directly from b.
+    
+    Returns:
+        row_masks (List[int]): Each integer is a bitmask representing one row of M.
+        parity (List[int]): Each element is 0 or 1, taken from b.
```
You should be specifying the return types via type hints for the function (not in the doc-string).

vprusso commented on toqito/nonlocal_games/binary_constraint_system_game.py:
```diff
@@ -0,0 +1,127 @@
+    # If there's a cycle, a perfect strategy exists
+    return bool(nx.cycle_basis(G))
+
+def nonlocal_game_from_constraints(constraints) -> NonlocalGame:
```
You need to type-hint the arguments for your functions.

vprusso commented on toqito/nonlocal_games/tests/test_binary_constraint_system_game.py:
```diff
@@ -0,0 +1,53 @@
+    # Expecting a perfect strategy => True
+    assert check_perfect_commuting_strategy(M, b)
+    game = NonlocalGame.from_bcs_game(constraints, reps=1)
+    #assert game.has_perfect_commuting_measurement_strategy()
```
Did you resolve the comments from before? I believe I previously asked about why these are commented out.

Wilson123d replied:
> **Note:** I created a new file called `modified_nonlocal_game.py`, which is a duplicate of `nonlocal_game.py`.
> 
> I only changed the `from_bcs_game` method inside the `NonlocalGame` class for testing.
> 
> Everything should now be consistent. If this change looks good, I’ll delete `modified_nonlocal_game.py` and apply the final changes directly in `nonlocal_game.py`.
>
>  **All tests are now passing.**

vprusso commented on toqito/nonlocal_games/binary_constraint_system_game.py:
```diff
@@ -0,0 +1,184 @@
+    --------
+    .. code-block:: python
+
+        import numpy as np
```
Please refer to other docstrings in toqito to see how we represent examples.

vprusso commented on toqito/nonlocal_games/binary_constraint_system_game.py:
```diff
@@ -0,0 +1,184 @@
+       A cycle implies a nontrivial group element, which corresponds
+       to the existence of a perfect strategy.
+
+    Examples
+    --------
```
Please refer to the toqito docs to see how represent the subject headings.

vprusso commented on toqito/nonlocal_games/binary_constraint_system_game.py:
```diff
@@ -0,0 +1,184 @@
+    Determine whether a perfect commuting-operator strategy exists for a BCS game.
+
+    This function checks if there is a perfect commuting-operator strategy
+    for the system of binary constraints ``M x = b``. The logic is based on
```
The phrase M x = b should not be represented by "code", it is a "math"-ematical statement (and should use the appropriate rendering environment.

vprusso commented on toqito/nonlocal_games/binary_constraint_system_game.py:
```diff
@@ -0,0 +1,184 @@
+    :return: A list of 1D NumPy arrays, where each array is of length ``n+1``.
+             The first ``n`` entries correspond to a row of ``M``, and the last
+             entry is :math:`\pm 1` depending on ``b[i]``.
+    :rtype: List[np.ndarray]
```
Please do not put in explicit return types. We know what these should be based on the type-hinting.

vprusso reviewed (commented):
Thanks for the updates, @Wilson123d 

I'm not sure I follow the `modified_nonlocal_game.py` approach, but if you think that helps you in scaffolding things, that's okay as a temporary thing.

purva-thakre commented:
@Wilson123d What's the status on this?

Wilson123d commented:
> @Wilson123d What's the status on this?

> Thanks for your message!

Last time, I modified some design recipes for the function and updated the `from_bcs_game` method inside the `NonlocalGame` class for testing. All tests were passing at that point.

I also created a new file called `modified_nonlocal_game.py`, which is currently a duplicate of `nonlocal_game.py`.

If we're going to modify `nonlocal_game.py` directly, then I can delete `modified_nonlocal_game.py`.

vprusso commented:
Hi @Wilson123d ,

Yes, ideally we would be modifying the existing file to include this logic instead of having some auxiliary file. If you could proceed with the MR, we can continue to offer our feedback if helpful. Thank you, @Wilson123d

vprusso commented:
Hi @Wilson123d 

Are you still intending on working on this MR? It seems like there are still some outstanding comments, conflicts, etc. Thank you!

Wilson123d commented:
> Hi @Wilson123d
> 
> Are you still intending on working on this MR? It seems like there are still some outstanding comments, conflicts, etc. Thank you!

Hi @vprusso, thanks for the reminder!

> I originally thought the MR was working fine apart from the `RuntimeWarning` related to `dependent_variables`.

> After digging deeper, I realized that some constraints could result in uniform tensors (all ones), which caused `np.diff(..., axis=i).any()` to return `False` across all axes — leading to a division-by-zero in `p_y = dependent_variables[j] / num_dependent_vars`.

> I’ve now fixed this by initializing each constraint tensor with:
> ```python
> constraint = np.full((2,) * n, fill_value=-rhs, dtype=int)
> constraint[idx] = rhs
> ```
> This guarantees variation in the tensor and ensures `dependent_variables` are detected correctly.

All tests now pass without warnings (`test_magic_square_bcs`, `test_special_case`, etc.), and I’ll push the updated changes and resolve the remaining comments shortly.

Thanks again for your patience!

vprusso commented on toqito/nonlocal_games/nonlocal_game.py:
```diff
@@ -81,6 +84,29 @@ def from_bcs_game(cls, constraints: list[np.ndarray], reps: int = 1) -> "Nonloca
         :param reps: Number of parallel repetitions to perform. Default is 1.
         :return: A NonlocalGame object arising from the variables and constraints that define the game.
         """
+        # Detect if constraints are in 1D form (each a vector of length n+1).
```
I don't think the placement of this makes sense. In the event that the user wants to invoke `from_bcs`, we don't want to simultaneously also do a check to see if the strategy is perfect. Instead, I would anticipate a separate class method in the `NonlocalGame` class called `is_bcs_perfect_commuting_strategy`. This function can then be conditionally invoked on the `NonlocalGame` object to return a boolean. 

Currently, there is too much conflation between the two items and this makes the computation for conversion longer as well as includes an interface for BCS that goes against the interface for all other NLG objects. Does this make sense to you?

Wilson123d replied:
>>> @vprusso Thanks for pointing this out. I have revised the design so that `from_bcs_game` no longer attempts to detect whether a perfect commuting‑operator strategy exists during the game’s construction. 

I introduced a separate instance method on `NonlocalGame`:

```python
def is_bcs_perfect_commuting_strategy(self) -> bool:
    ...
This method can be invoked on the constructed 'NonlocalGame' object at any time to check for a perfect commuting‑operator strategy. It handles the conversion of constraints (if necessary) into the form required by 'check_perfect_commuting_strategy', but it no longer conflates game building with strategy checking.

Wilson123d replied:
> @vprusso Thanks for pointing this out. I have revised the design so that `from_bcs_game` no longer attempts to detect whether a perfect commuting‑operator strategy exists during the game’s construction. 

I introduced a separate instance method on `NonlocalGame`:

```python
def is_bcs_perfect_commuting_strategy(self) -> bool:
    ...
```
This method can be invoked on the constructed `NonlocalGame` object at any time to check for a perfect commuting‑operator strategy. It handles the conversion of constraints (if necessary) into the form required by `check_perfect_commuting_strategy`, but it no longer conflates game building with strategy checking.

purva-thakre commented:
@Wilson123d Is this PR ready for a review?

Wilson123d commented:
Thanks for your message. Yes I think so. I included it in the submission of proposal. Thanks!

Best,
Yangchen Zhou
________________________________
发件人: Purva Thakre ***@***.***>
发送时间: 2025年5月7日 3:59
收件人: vprusso/toqito ***@***.***>
抄送: Yangchen Zhou ***@***.***>; Mention ***@***.***>
主题: Re: [vprusso/toqito] Add Group-Theoretic Verification for Perfect Commuting Strategies in BCS Games (PR #1043)

[https://avatars.githubusercontent.com/u/66048318?s=20&v=4]purva-thakre left a comment (vprusso/toqito#1043)<https://github.com/vprusso/toqito/pull/1043#issuecomment-2855787041>

@Wilson123d<https://github.com/Wilson123d> Is this PR ready for a review?

―
Reply to this email directly, view it on GitHub<https://github.com/vprusso/toqito/pull/1043#issuecomment-2855787041>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BBR55ISBY52LXZHVTLE6DKL25EIC7AVCNFSM6AAAAABZPMI6ICVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNJVG44DOMBUGE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>
